hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java

```diff
@@ -661,6 +652,35 @@ private Pair<SchemaProvider, Pair<String, JavaRDD<HoodieRecord>>> fetchFromSourc
     return Pair.of(schemaProvider, Pair.of(checkpointStr, records));
   }
 
+  /**
+   * Apply schema reconcile and schema evolution rules(schema on read) and generate new target schema provider.
+   *
+   * @param incomingSchema schema of the source data
+   * @param sourceSchemaProvider Source schema provider.
+   * @return the SchemaProvider that can be used as writer schema.
+   */
+  private SchemaProvider getDeducedSchemaProvider(Schema incomingSchema, SchemaProvider sourceSchemaProvider) {
```

Comment:
There's another method in this class `getSchemaForWriteConfig` that does something that looks similar. Is that now obsolete?

Reply:
Yeah, I am confused about this

Reply:
Let's try to step through a case that can hit that path today.

Reply:
This function just picks up latest table schema for writing in case schema provider is set to NULL schema. All the evolution is handled in getDeducedSchemaProvider api